hollow/src/main/java/com/netflix/hollow/api/consumer/metrics/AbstractRefreshMetricsListener.java

```diff
@@ -0,0 +1,178 @@
+package com.netflix.hollow.api.consumer.metrics;
+
+import com.netflix.hollow.api.consumer.HollowConsumer;
+import com.netflix.hollow.api.custom.HollowAPI;
+import com.netflix.hollow.core.read.engine.HollowReadStateEngine;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+
+public class AbstractRefreshMetricsListener extends HollowConsumer.AbstractRefreshListener  {
+
+    private long consecutiveSuccesses;
+    private long consecutiveFailures;
+
+    private Instant lastRefreshInstant;
+    private Instant refreshStart;
+
+    private ConsumerRefreshStatus refreshStatus;                    // success or failure
+    private ConsumerRefreshTransitionType refreshTransitionType;    // snapshot or delta transition
+    private ConsumerRefreshLoadType refreshLoadType;                // "init" indicates first successful load on consumer, "subsequent" indicates subsequent load during refresh
+
+    private long refreshLoadCount;
+
+    // Encapsulate the refresh metrics so that more metrics/properties can be added in future wihtout breaking existing clients.
+    public static class ConsumerRefreshMetrics {
+        private long refreshDurationMillis;
+        private ConsumerRefreshStatus refreshStatus;
+        private ConsumerRefreshTransitionType refreshTransitionType;
+        private ConsumerRefreshLoadType refreshLoadType;
+
+        public ConsumerRefreshMetrics(long refreshDurationMillis, ConsumerRefreshStatus refreshStatus, ConsumerRefreshTransitionType refreshTransitionType, ConsumerRefreshLoadType refreshLoadType) {
+            this.refreshDurationMillis = refreshDurationMillis;
+            this.refreshStatus = Objects.requireNonNull(refreshStatus);
+            this.refreshTransitionType = refreshTransitionType;   // Transition type can be unknown for some failed snapshot or delta transition
+            this.refreshLoadType = Objects.requireNonNull(refreshLoadType);
+        }
+
+        public long getRefreshDurationMillis() {
+            return refreshDurationMillis;
+        }
+        public ConsumerRefreshStatus getRefreshStatus() {
+            return refreshStatus;
+        }
+        public ConsumerRefreshTransitionType getRefreshTransitionType() {
```

Comment:
That sounds like a good strategy. 

On one hand, it feels like having that finer grain visibility is useful, and on the other hand, since individual _transitions_ in a multi-_transition_ update plan don't provide fresh data to the consumer (for the lack of a better description) it's useful to also look at the refresh as a whole.

Currently, snapshotApplied() and deltaApplied() give good visibility on the individual _transitions_ , but the behavior of snapshotUpdateOccurred() and deltaUpdateOccurred() is-

For a single refresh involving snapshot only or snapshot + n deltas,
snapshotUpdateOccurred() is currently called once at the end.
For a single refresh involving n deltas,
deltaUpdateOccurred() is currently called once after every delta transition.

Re-working deltaUpdateOccurred() might solve our problems here, but not sure what side-effects that might lead to. Lets chat about this further to validate that my understand is correct, and get your thoughts on it.

Reply:
overallRefreshType now embodies 3 types-
"snapshot", for 1 snapshot or 1 snapshot + N deltas
"delta", for N deltas
"reversedelta" for N reverse deltas
As per your comment, I didn't want to make this a standard definition for an update plan, so it's evaluated in the Metrics listener. Thought it might be okay to bias with this definition for refresh metrics purposes.

I also now have a list of transition types comprising a refresh, for eg, [SNAPSHOT, DELTA, DELTA] that might be useful for logging for failure scenarios. More detail can be added later, thought we could start with this much.